test/functional/testlib.bash

```diff
@@ -0,0 +1,502 @@
+	# Every bundle has to have at least one directory,
+	# hashes in directories vary depending on owner and permissions,
+	# so one directory hash can be reused many times
+	bundle_dir=$(create_dir "$files_path")
+	echo "Directory -> $bundle_dir"  # TODO(castulo): remove this msg when finish development
+	# Create a zero pack for the bundle and add the directory to it
+	sudo tar -cf "$version_path"/pack-"$bundle_name"-from-0.tar --exclude="$bundle_dir"/*  "$bundle_dir"
+	for val in "${dir_list[@]}"; do
+		add_to_manifest "$manifest" "$bundle_dir" "$val"
+		if [ "$local_bundle" = true ]; then
+			sudo mkdir -p "$target_path$val"
+		fi
+	done
+	
+	# 3) Create the requested file(s)
+	for val in "${file_list[@]}"; do
+		# if the directories the file is don't exist, add them to the
+		# bundle (except for the dir of the tracking file)
+		fdir=$(dirname "$val")
+		if [ ! "$(sudo cat "$manifest" | grep -x "D\\.\\.\\..*$fdir")" ] && [ "$fdir" != "/usr/share/clear/bundles" ]; then
+			bundle_dir=$(create_dir "$files_path")
+			add_to_manifest "$manifest" "$bundle_dir" "$fdir"
+			# add each one of the directories of the path if they are not in the manifest already
+			while [ $(dirname "$fdir") != "/" ]; do
+				fdir=$(dirname "$fdir")
+				if [ ! "$(sudo cat "$manifest" | grep -x "D\\.\\.\\..*$fdir")" ]; then
+					add_to_manifest "$manifest" "$bundle_dir" "$fdir"
+				fi
+			done
+		fi
+		bundle_file=$(create_file "$files_path")
+		echo "File -> $bundle_file"  # TODO(castulo): remove this msg when finish development
+		add_to_manifest "$manifest" "$bundle_file" "$val"
+		# Add the file to the zero pack of the bundle
+		sudo tar -rf "$version_path"/pack-"$bundle_name"-from-0.tar "$bundle_file"
+		# if the local_bundle flag is set, copy the files to the target-dir as if the
+		# bundle had been locally installed
+		if [ "$local_bundle" = true ]; then
+			sudo mkdir -p "$target_path$(dirname "$val")"
+			sudo cp "$bundle_file" "$target_path$val"
+		fi 
+	done
+	
+	# 4) Create the requested link(s) in the bundle
+	for val in "${link_list[@]}"; do
+		# if the directory the link is doesn't exist,
+		# add it to the bundle
+		fdir=$(dirname "$val")
+		if [ ! "$(sudo cat "$manifest" | grep -x "D\\.\\.\\..*$fdir")" ]; then
+			bundle_dir=$(create_dir "$files_path")
+			add_to_manifest "$manifest" "$bundle_dir" "$fdir"
+		fi
+		bundle_link=$(create_link "$files_path")
+		echo "Symlink -> $bundle_link"  # TODO(castulo): remove this msg when finish development
+		add_to_manifest "$manifest" "$bundle_link" "$val"
+		# Add the file pointed by the link to the zero pack of the bundle
+		pfile=$(readlink -f "$bundle_link")
+		sudo tar -rf "$version_path"/pack-"$bundle_name"-from-0.tar "$pfile"
+		if [ "$local_bundle" = true ]; then
+			sudo mkdir -p "$target_path$(dirname "$val")"
+			# if local_bundle is enabled copy the link to target-dir but also
+			# copy the file it points to
+			pfile_path=$(sudo cat "$manifest" | grep "$(basename "$pfile")" | awk '{ print $4 }')
```

Comment:
`grep | awk` can always be replaced by `awk` (although it might be less readable). Almost always just having awk will be faster.

Reply:
done